### PR TITLE
IDVA6-1189: Remove Company

### PR DESCRIPTION
**Pull request by kwamenyantakyi-ch**
**Link To Tickets:**
https://companieshouse.atlassian.net/browse/IDVA6-1189
https://companieshouse.atlassian.net/browse/IDVA6-1250
https://companieshouse.atlassian.net/browse/IDVA6-1251

**Changes Made:**
- Implemented company status, remove company link on your companies page.
- Implemented remove-company.njk page and controllers for the page.
- Implemented remove company success page and controllers for the page.
- Implemented unit tests to test the controllers -> YourCompaniesController, RemoveCompanyController, RemoveCompanyConfrimed.

**Screenshots:**
Screenshot 2024-09-30 at 13 22 51
Screenshot 2024-09-30 at 13 23 04
Screenshot 2024-09-30 at 13 23 12

### CODE REVIEW

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/routers/handlers/removeCompanyHandler.ts [hunk 1]
```diff
@@ -0,0 +1,98 @@
+import { Request, Response } from "express";
```
Comment (mmodzelewski-ch) [inline src/routers/handlers/removeCompanyHandler.ts:1]: This file is at the wrong level. It should be inside yourCompanies folder, so the path should be src/routers/handlers/yourCompanies/removeCompanyHandler.ts

Also it has very low test coverage. The red numbers indicate lines of code not covered by unit tests at all
Screenshot 2024-10-03 at 11 24 23

File: src/routers/controllers/removeCompanyController.ts [hunk 2]
```diff
@@ -0,0 +1,32 @@
+import { Request, Response } from "express";
```
Comment (mmodzelewski-ch) [inline src/routers/controllers/removeCompanyController.ts:1]: This controller has low test coverage. The red numbers indicate lines of code not covered by tests at all.
Screenshot 2024-10-03 at 11 24 54

**Review by mmodzelewski-ch — CHANGES_REQUESTED** [review]

File: src/constants.ts [hunk 3]
```diff
@@ -103,7 +109,8 @@ export const CONFIRMATION_PERSON_REMOVED_URL = "/confirmation-person-removed";
 export const CONFIRMATION_PERSON_ADDED = "/confirmation-person-added";
 export const MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${MANAGE_AUTHORISED_PEOPLE_URL}${CONFIRMATION_PERSON_REMOVED_URL}`;
 export const YOUR_COMPANIES_MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${LANDING_URL}${MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL}`;
-
+export const YOUR_COMPANIES_REMOVE_COMPANY_URL = `${LANDING_URL}/remove-company/:${COMPANY_NUMBER}`;
+export const REMOVE_COMPANY_URL = `/remove-company/:${COMPANY_NUMBER}`;
```
[the remark this answers is not in the document]
Comment (tferns-ch) [reply src/constants.ts]: Good eye, we've refactored this to use the constant in https://github.com/companieshouse/your-companies-web/pull/191/commits/9a9b84b1dba84c7ddadfb555e459070cd6e2ff6e

File: src/constants.ts [hunk 4]
```diff
@@ -103,7 +109,8 @@ export const CONFIRMATION_PERSON_REMOVED_URL = "/confirmation-person-removed";
 export const CONFIRMATION_PERSON_ADDED = "/confirmation-person-added";
 export const MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${MANAGE_AUTHORISED_PEOPLE_URL}${CONFIRMATION_PERSON_REMOVED_URL}`;
 export const YOUR_COMPANIES_MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${LANDING_URL}${MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL}`;
-
+export const YOUR_COMPANIES_REMOVE_COMPANY_URL = `${LANDING_URL}/remove-company/:${COMPANY_NUMBER}`;
```
[the remark this answers is not in the document]
Comment (tferns-ch) [reply src/constants.ts]: Good eye, we done this too in https://github.com/companieshouse/your-companies-web/pull/191/commits/9a9b84b1dba84c7ddadfb555e459070cd6e2ff6e

File: src/lib/utils/sessionUtils.ts [hunk 5]
```diff
@@ -34,3 +34,10 @@ export const deleteExtraData = (session: Session | undefined, key: string): bool
 export const getAccessToken = (session: Session | undefined): string => {
     return session?.data.signin_info?.access_token?.access_token as string;
 };
+
+export const popExtraData = (session: Session | undefined, key: string): any => {
```
[the remark this answers is not in the document]
Comment (tferns-ch) [reply src/lib/utils/sessionUtils.ts:38]: Added unit tests in https://github.com/companieshouse/your-companies-web/blob/9a9b84b1dba84c7ddadfb555e459070cd6e2ff6e/test/src/lib/utils/sessionUtils.test.ts - good idea

File: src/routers/handlers/removeCompanyHandler.ts (hunk 1 above)
Comment (tferns-ch) [reply src/routers/handlers/removeCompanyHandler.ts]: We improved this coverage by refactoring the code to get rid of unnecessary errors, what do you think about it now?

File: src/routers/controllers/removeCompanyController.ts (hunk 2 above)
Comment (tferns-ch) [reply src/routers/controllers/removeCompanyController.ts:1]: Similar to above, we refactored to get rid unnecessary lines, could you re-review please?

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

File: src/constants.ts [hunk 6]
```diff
@@ -103,7 +109,8 @@ export const CONFIRMATION_PERSON_REMOVED_URL = "/confirmation-person-removed";
 export const CONFIRMATION_PERSON_ADDED = "/confirmation-person-added";
 export const MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${MANAGE_AUTHORISED_PEOPLE_URL}${CONFIRMATION_PERSON_REMOVED_URL}`;
 export const YOUR_COMPANIES_MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL = `${LANDING_URL}${MANAGE_AUTHORISED_PEOPLE_CONFIRMATION_PERSON_REMOVED_URL}`;
-
+export const REMOVE_COMPANY_URL = `/${REMOVE_COMPANY_PAGE}/:${COMPANY_NUMBER}`;
+export const YOUR_COMPANIES_REMOVE_COMPANY_URL = `${LANDING_URL}/${REMOVE_COMPANY_PAGE}/:${COMPANY_NUMBER}`;
```
Comment (mmodzelewski-ch) [inline src/constants.ts:113]: Why not `${LANDING_URL}${REMOVE_COMPANY_URL}` ?

File: src/routers/handlers/removeCompanyHandler.ts (hunk 1 above)
Comment (mmodzelewski-ch) [reply src/routers/handlers/removeCompanyHandler.ts]: SonarQube indicates that the parts of the code that deal with errors are not covered with tests at all
Screenshot 2024-10-03 at 16 27 02

**Review by mmodzelewski-ch — CHANGES_REQUESTED** [review]

File: src/constants.ts (hunk 6 above)
Comment (tferns-ch) [reply src/constants.ts:113]: 🤦 this went over our heads, nice spot

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by mmodzelewski-ch — CHANGES_REQUESTED** [review]
See SonarQube for the code smell and the one file with zero test coverage

**Comment by ch-code-analysis** [comment]
CI: No security warnings found

**Review by mmodzelewski-ch — APPROVED** [review]

**Pull request merged**